haskoin-core/src/Network/Haskoin/Block/Types.hs

```diff
@@ -339,28 +335,51 @@ instance Serialize Headers where
 -- Bit number 24 (0x800000) represents the sign of N.
 --
 -- >    N = (-1^sign) * mantissa * 256^(exponent-3)
-decodeCompact :: Word32 -> Integer
-decodeCompact c =
-    if neg then (-res) else res
+decodeCompact :: Word32 -> (Integer, Bool) -- ^ overflow
+decodeCompact nCompact = (if neg then res * (-1) else res, over)
   where
-    size = fromIntegral $ c `shiftR` 24
-    neg  = (c .&. 0x00800000) /= 0
-    wrd  = c .&. 0x007fffff
-    res | size <= 3 = toInteger wrd `shiftR` (8*(3 - size))
-        | otherwise = toInteger wrd `shiftL` (8*(size - 3))
+    nSize :: Int
+    nSize = fromIntegral nCompact `shiftR` 24
+    nWord' :: Word32
+    nWord' = nCompact .&. 0x007fffff
+    nWord :: Word32
+    nWord | nSize <= 3 = nWord' `shiftR` (8 * (3 - nSize))
+          | otherwise = nWord'
+    res :: Integer
+    res | nSize <= 3 = fromIntegral nWord
+        | otherwise = fromIntegral nWord `shiftL` (8 * (nSize - 3))
+    neg = nWord /= 0 && (nCompact .&. 0x00800000) /= 0
+    over = nWord /= 0 && (nSize > 34 ||
+                          nWord > 0xff && nSize > 33 ||
+                          nWord > 0xffff && nSize > 32)
 
 -- | Encode an Integer to the compact number format used in the difficulty
 -- target of a block.
-encodeCompact :: Integer -> Word32
-encodeCompact i
-    | i < 0     = c3 .|. 0x00800000
-    | otherwise = c3
+encodeCompact :: Integer
+              -> Word32
+encodeCompact i = nCompact
   where
-    posi = abs i
-    s1 = BS.length $ integerToBS posi
-    c1 | s1 < 3    = posi `shiftL` (8*(3 - s1))
-       | otherwise = posi `shiftR` (8*(s1 - 3))
-    (s2,c2) | c1 .&. 0x00800000 /= 0  = (s1 + 1, c1 `shiftR` 8)
-            | otherwise               = (s1, c1)
-    c3 = fromIntegral $ c2 .|. (toInteger s2 `shiftL` 24)
-
+    i' = abs i
+    neg = i < 0
+    nSize' :: Int
+    nSize' = let f n = let n' = n `shiftR` 8
```

Comment:
According to https://github.com/bitcoin/bitcoin/blob/master/src/arith_uint256.cpp#L228 `nSize == 0` is a valid possibility but this code will always give at least `1`. Maybe consider something like this:

```
nSize' = let f n = if n == 0 then 0 else 1 + f (n `shiftR` 8)
          in f i' 
```

Reply:
That was an excellent catch sir.

![Image of a big fish](http://vermilionbaylodge.com/wp-content/gallery/ice-fishing-with-chris-nick_1/tn_chunky-laker.jpg)

I added new test vectors that fail with the buggy code, as well as a fix for this issue.